### PR TITLE
Add in the GpuArrayFilter command

### LINKED ISSUE #10760: [FEA]Support ArrayFilter
I wish we can support ArrayFilter.
eg:
```
scala>  spark.sql("SELECT filter(array(1, 2, 3), x -> x % 2 == 1)").show
```

Not-supported-messages:
```
      ! <ArrayFilter> filter([1,2,3], lambdafunction(((lambda x#0 % 2) = 1), lambda x#0, false)) cannot run on GPU because GPU does not currently support the operator class org.apache.spark.sql.catalyst.expressions.ArrayFilter
```

### PR DESCRIPTION
**Pull request by revans2**
this fixes #10760

I still need to do some performance comparisons/testing, but functionality wise it looks good.  Will update it shortly.

### CODE REVIEW

**Comment by revans2** [comment]
build

**Comment by revans2** [comment]
For performance numbers I ran the following query on Spark 3.4.2

```
spark.time(spark.range(0, 10000000000L, 1, 512).selectExpr("array(id, id + 1, id - 100, id + 3, id + 2) as ar").selectExpr("filter(ar, f -> f % 3 == 0) as f").selectExpr("SUM(size(f)) as r").show())
```

On an a6000 running with 4 concurrent It took 36 seconds in the worst case, I ran it 3 times and the best was about 33 seconds.

On a 32-core/64-thread Threadripper PRO 5975WX, using all 64 cores it took 688 seconds (and I didn't really want to run it again).

That is a speedup of about 19x.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/higherOrderFunctions.scala [hunk 1]
```diff
@@ -278,7 +278,8 @@ trait GpuArrayTransformBase extends GpuSimpleHigherOrderFunction {
    * Post-process the column view of the array after applying the function parameter
    */
   protected def transformListColumnView(
-    lambdaTransformedCV: cudf.ColumnView): GpuColumnVector
+    lambdaTransformedCV: cudf.ColumnView,
+    arg: cudf.ColumnView): GpuColumnVector
```
Comment (abellina) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/higherOrderFunctions.scala:282]: would be nice to add the definition of `arg` in this method header, it's not as clear to me as `lambdaTransformedCV`

**Comment by revans2** [comment]
build

File: integration_tests/src/main/python/array_test.py
Comment (jlowe) [inline integration_tests/src/main/python/array_test.py:1]: 2024 copyrights

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/higherOrderFunctions.scala
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/higherOrderFunctions.scala:1]: 2024 copyrights

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/higherOrderFunctions.scala [hunk 2]
```diff
@@ -336,9 +339,9 @@ case class GpuArrayExists(
   }
 
   private def imputeFalseForEmptyArrays(
-    transformedCV: cudf.ColumnView,
-    result: cudf.ColumnView
-  ): GpuColumnVector = {
+                                         transformedCV: cudf.ColumnView,
+                                         result: cudf.ColumnView
+                                       ): GpuColumnVector = {
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/higherOrderFunctions.scala:344]: Not sure why this and `transformListColumnView` indentation of arguments was changed here?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/higherOrderFunctions.scala [hunk 3]
```diff
@@ -402,8 +405,9 @@ case class GpuArrayExists(
   }
 
   override protected def transformListColumnView(
-    lambdaTransformedCV: cudf.ColumnView
-  ): GpuColumnVector = {
+                                                  lambdaTransformedCV: cudf.ColumnView,
+                                                  arg: cudf.ColumnView
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/higherOrderFunctions.scala:409]: Same indentation question here, not consisten with ident used for `GpuArrayFilter` args.

**Review by abellina — DISMISSED** [review]
LGTM other than copyright/indenting that @jlowe pointed out.

**Comment by revans2** [comment]
@jlowe @abellina please take another look

**Comment by revans2** [comment]
build

**Comment by revans2** [comment]
The link check is failing because of https://github.com/NVIDIA/spark-rapids/issues/10768

**Review by jlowe — APPROVED** [review]

**Review by abellina — APPROVED** [review]

**Pull request merged**